lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,832 @@
+  template<typename A> bool PushNonBlockName(const A& X) {
+    const auto& OptName{std::get<0>(std::get<0>(X.t).statement.t)};
+    if (OptName.has_value())
+      Names.push_back(OptName.value().ToString());
+    return true;
+  }
+
+  template<typename A> void PopNonBlockConstructName(const A& X) {
+    CheckName(X); SelectivePopBack(X);
+  }
+
+  template<typename A> void SelectivePopBack(const A& X) {
+    const auto& OptName{std::get<0>(std::get<0>(X.t).statement.t)};
+    if (OptName.has_value())
+      Names.pop_back();
+  }
+  void SelectivePopBack(const parser::BlockConstruct& Blk) {
+    const auto& OptName{std::get<0>(Blk.t).statement.v};
+    if (OptName.has_value())
+      Names.pop_back();
+  }
+
+  /// \brief Check constraints and pop scope
+  template<typename A> void PopName(const A& V) {
+    CheckName(V); PopScope(); SelectivePopBack(V);
+  }
+
+  /// \brief Check <i>case-construct-name</i> and pop the scope
+  /// Constraint C1144 - opening and ending name must match if present, and
+  /// <i>case-stmt</i> must either match or be unnamed
+  void PopName(const parser::CaseConstruct& Case) {
+    CheckName(Case, "CASE"); PopScope(); SelectivePopBack(Case);
+  }
+
+  /// \brief Check <i>select-rank-construct-name</i> and pop the scope
+  /// Constraints C1154, C1156 - opening and ending name must match if present,
+  /// and <i>select-rank-case-stmt</i> must either match or be unnamed
+  void PopName(const parser::SelectRankConstruct& SelRk) {
+    CheckName(SelRk, "RANK","RANK "); PopScope(); SelectivePopBack(SelRk);
+  }
+
+  /// \brief Check <i>select-construct-name</i> and pop the scope
+  /// Constraint C1165 - opening and ending name must match if present, and
+  /// <i>type-guard-stmt</i> must either match or be unnamed
+  void PopName(const parser::SelectTypeConstruct& SelTy) {
+    CheckName(SelTy, "TYPE", "TYPE "); PopScope(); SelectivePopBack(SelTy);
+  }
+
+  // -----------------------------------------------
+  // CheckName - check constraints on construct-name
+  // Case 1: construct name must be absent or specified & identical on END
+
+  /// \brief Check <i>associate-construct-name</i>, constraint C1106
+  void CheckName(const parser::AssociateConstruct& A) { ChkNm(A, "ASSOCIATE"); }
+  /// \brief Check <i>critical-construct-name</i>, constraint C1117
+  void CheckName(const parser::CriticalConstruct& C) { ChkNm(C, "CRITICAL"); }
+  /// \brief Check <i>do-construct-name</i>, constraint C1131
+  void CheckName(const parser::DoConstruct& Do) { ChkNm(Do, "DO"); }
+  /// \brief Check <i>forall-construct-name</i>, constraint C1035
+  void CheckName(const parser::ForallConstruct& F) { ChkNm(F, "FORALL"); }
+  /// \brief Common code for ASSOCIATE, CRITICAL, DO, and FORALL
+  template<typename A> void ChkNm(const A& V, const char *const Con) {
+    if (!BothEqOrNone(std::get<0>(std::get<0>(V.t).statement.t),
+		      std::get<2>(V.t).statement.v)) {
+      EH.Report(Index, "%s construct name mismatch"_err_en_US, Con);
```

Comment:
You can attach an information message to the error that points to the name you are trying to match. See `Say2` in `resolve-names.cc` for an example. At minimum I think the message should include both names.

Reply:
Right. I didn't spend very much time trying to craft "supermodel" error messages. I was thinking of doing that as another refactoring pass.